File: examples/fortran/example_netcdf4_write.f90
!> @file example_netcdf4_write.f90
!> @brief Example: Write a 3D temperature field to NetCDF-4 using AMIO (Fortran).
!>
!> This example demonstrates the complete AMIO lifecycle from Fortran for
!> writing a 3D atmospheric temperature field (100 x 50 x 25: lon x lat x level)
!> to a NetCDF-4 file using the netcdf4 backend driver.
!>
!> Steps:
!>   1. Initialize AMIO with a manifest targeting the netcdf4 backend
!>   2. Open a dataset for writing
!>   3. Populate a 3D real(4) array with synthetic temperature data
!>   4. Write the array using the amio_write_array convenience subroutine
!>   5. Flush to ensure all writes complete
!>   6. Close the dataset and finalize AMIO
!>
!> Build (via CMake):
!>   Linked against AMIO::amio_fortran
!>
!> Run:
!>   ./example_netcdf4_write_f90

program example_netcdf4_write
    use, intrinsic :: iso_c_binding, only: &
        c_ptr, c_null_ptr, c_int32_t, c_int64_t, c_float, c_char, c_null_char
    use amio_mod
    implicit none

    ! Grid dimensions
    integer, parameter :: NX = 100   ! longitude points
    integer, parameter :: NY = 50    ! latitude points
    integer, parameter :: NZ = 25    ! vertical levels

    ! AMIO handles
    type(c_ptr) :: core
    type(c_ptr) :: dataset
    type(c_ptr) :: io_handle
    integer(c_int32_t) :: rc

    ! Data array
    real(c_float), allocatable, target :: temperature(:, :, :)

    ! Local variables
    integer :: i, j, k
    real(c_float) :: lat, lon, altitude_km, base_temp, temp
    real(c_float), parameter :: PI = 3.14159265_c_float

    write (*, '(A)') 'AMIO NetCDF-4 Write Example (Fortran)'
    write (*, '(A)') '======================================'
    write (*, '(A,I0,A,I0,A,I0,A)') &
        'Writing 3D temperature field (', NX, ' x ', NY, ' x ', NZ, ') to NetCDF-4'
    write (*, *)

    ! -------------------------------------------------------------------
    ! Step 1: Initialize AMIO with the netcdf4 manifest.
    !
    ! Note: Fortran strings passed to C must be null-terminated using
    ! c_null_char. The amio_init function expects a C string.
    ! -------------------------------------------------------------------
    write (*, '(A)') 'Step 1: Initializing AMIO...'
    rc = amio_init('examples/manifests/netcdf4_manifest.yaml'//c_null_char, core)
    call check_amio(rc, 'amio_init')
    write (*, '(A)') '  AMIO initialized successfully.'

    ! -------------------------------------------------------------------
    ! Step 2: Open a dataset for writing.
    !
    ! AMIO_MODE_WRITE creates a new file (or overwrites if configured).
    ! -------------------------------------------------------------------
    write (*, '(A)') 'Step 2: Opening dataset for writing...'
    rc = amio_open_dataset(core, &
                           'examples/manifests/netcdf4_manifest.yaml'//c_null_char, &
                           AMIO_MODE_WRITE, dataset)
    call check_amio(rc, 'amio_open_dataset')
    write (*, '(A)') '  Dataset opened.'

    ! -------------------------------------------------------------------
    ! Step 3: Allocate and fill the temperature array.
    !
    ! Generate a realistic temperature field that decreases with altitude
    ! (standard atmosphere lapse rate) and varies with latitude.
    ! -------------------------------------------------------------------
    write (*, '(A)') 'Step 3: Generating synthetic temperature data...'
    allocate (temperature(NX, NY, NZ))

    do k = 1, NZ
        altitude_km = real(k - 1, c_float) * 0.5_c_float
        do j = 1, NY
            lat = -90.0_c_float + real(j - 1, c_float) * (180.0_c_float / real(NY - 1, c_float))
            base_temp = 288.15_c_float - 30.0_c_float * abs(lat) / 90.0_c_float
            do i = 1, NX
                lon = real(i - 1, c_float) * (360.0_c_float / real(NX, c_float))
                temp = base_temp - 6.5_c_float * altitude_km
                temp = temp + 2.0_c_float * sin(lon * PI / 180.0_c_float)
                temperature(i, j, k) = temp
            end do
        end do
    end do

    write (*, '(A,F6.1,A,F6.1,A)') '  Data generated: T range ~ [', &
        minval(temperature), ', ', maxval(temperature), '] K'

    ! -------------------------------------------------------------------
    ! Step 4: Write the 3D array using the convenience subroutine.
    !
    ! amio_write_array is a generic interface that automatically:
    !   - Determines the data type (AMIO_DTYPE_F32 for real(c_float))
    !   - Computes the rank (3 for a 3D array)
    !   - Extracts extents from the array shape
    !   - Obtains the data pointer via c_loc()
    !   - Calls the raw amio_write C function
    !
    ! This is much simpler than manually constructing amio_shape_t.
    ! -------------------------------------------------------------------
    write (*, '(A)') 'Step 4: Writing temperature field...'
    call amio_write_f32_3d(dataset, 'temperature'//c_null_char, &
                           temperature, io_handle, rc)
    call check_amio(rc, 'amio_write_array(temperature)')
    write (*, '(A)') '  Write enqueued (async I/O handle obtained).'

    ! -------------------------------------------------------------------
    ! Step 5: Flush to ensure all pending writes complete.
    !
    ! timeout_ms = 0 means wait indefinitely.
    ! -------------------------------------------------------------------
    write (*, '(A)') 'Step 5: Flushing pending writes...'
    rc = amio_flush(dataset, 0_c_int64_t)
    call check_amio(rc, 'amio_flush')
    write (*, '(A)') '  All writes completed successfully.'

    ! -------------------------------------------------------------------
    ! Step 6: Close the dataset and finalize AMIO.
    ! -------------------------------------------------------------------
    write (*, '(A)') 'Step 6: Closing dataset and finalizing...'
    rc = amio_close_dataset(dataset)
    call check_amio(rc, 'amio_close_dataset')

    rc = amio_finalize(core)
    call check_amio(rc, 'amio_finalize')
    write (*, '(A)') '  AMIO finalized.'

    ! Clean up
    deallocate (temperature)

    write (*, *)
    write (*, '(A)') 'Done! Temperature field written to NetCDF-4.'

contains

    !> @brief Check an AMIO return code and stop on failure.
    subroutine check_amio(status, context)
        integer(c_int32_t), intent(in) :: status
        character(len=*), intent(in) :: context

        if (status /= AMIO_OK) then
            write (*, '(A,A,A,I0,A)') 'AMIO error in ', context, &
                ' (code ', status, ')'
            error stop 'AMIO operation failed'
        end if
    end subroutine check_amio

end program example_netcdf4_write

File: examples/c/example_netcdf4_write.c
/**
 * @file example_netcdf4_write.c
 * @brief Example: Write a 3D temperature field to NetCDF-4 using AMIO.
 *
 * This example demonstrates the complete AMIO lifecycle for writing a
 * 3D atmospheric temperature field (100 x 50 x 25: lon x lat x level)
 * to a NetCDF-4 file using the netcdf4 backend driver.
 *
 * Steps:
 *   1. Initialize AMIO with a manifest targeting the netcdf4 backend
 *   2. Open a dataset for writing
 *   3. Populate a 3D float array with synthetic temperature data
 *   4. Write the array through AMIO's async pipeline
 *   5. Flush to ensure all writes complete
 *   6. Close the dataset and finalize AMIO
 *
 * Build:
 *   cc -o example_netcdf4_write example_netcdf4_write.c -lamio
 *
 * Run:
 *   ./example_netcdf4_write
 */

#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <math.h>

#include <amio/amio.h>

/* Grid dimensions: longitude x latitude x vertical levels */
#define NX 100   /* longitude points */
#define NY  50   /* latitude points */
#define NZ  25   /* vertical levels */

/**
 * @brief Check an AMIO return code and exit on failure.
 */
static void check_amio(amio_status_t rc, const char *context)
{
    if (rc != AMIO_OK) {
        fprintf(stderr, "AMIO error in %s: %s (code %d)\n",
                context, amio_strerror(rc), (int)rc);
        exit(EXIT_FAILURE);
    }
}

/**
 * @brief Generate synthetic temperature data.
 *
 * Creates a realistic-looking temperature field that decreases with
 * altitude (standard atmosphere lapse rate ~6.5 K/km) and varies
 * with latitude (warmer at equator, cooler at poles).
 */
static void fill_temperature(float *data, int nx, int ny, int nz)
{
    for (int k = 0; k < nz; k++) {
        /* Approximate pressure level altitude (km) */
        float altitude_km = (float)k * 0.5f;
        for (int j = 0; j < ny; j++) {
            /* Latitude from -90 to +90 */
            float lat = -90.0f + (float)j * (180.0f / (float)(ny - 1));
            /* Base temperature varies with latitude (warmer at equator) */
            float base_temp = 288.15f - 30.0f * fabsf(lat) / 90.0f;
            for (int i = 0; i < nx; i++) {
                /* Standard atmosphere lapse rate */
                float temp = base_temp - 6.5f * altitude_km;
                /* Small longitudinal variation */
                float lon = (float)i * (360.0f / (float)nx);
                temp += 2.0f * sinf(lon * 3.14159f / 180.0f);
                data[k * ny * nx + j * nx + i] = temp;
            }
        }
    }
}

int main(void)
{
    amio_status_t rc;
    amio_core_handle core = NULL;
    amio_dataset_handle dataset = NULL;
    amio_io_handle io = NULL;

    printf("AMIO NetCDF-4 Write Example\n");
    printf("===========================\n");
    printf("Writing 3D temperature field (%d x %d x %d) to NetCDF-4\n\n",
           NX, NY, NZ);

    /* ---------------------------------------------------------------
     * Step 1: Initialize AMIO with the netcdf4 manifest.
     *
     * The manifest specifies the backend driver, staging pool size,
     * worker thread count, and output path.
     * --------------------------------------------------------------- */
    printf("Step 1: Initializing AMIO...\n");
    rc = amio_init("examples/manifests/netcdf4_manifest.yaml", &core);
    check_amio(rc, "amio_init");
    printf("  AMIO initialized successfully.\n");

    /* ---------------------------------------------------------------
     * Step 2: Open a dataset for writing.
     *
     * The dataset configuration path can be the same manifest or a
     * separate dataset-specific config. AMIO_MODE_WRITE creates a
     * new file (or overwrites if configured).
     * --------------------------------------------------------------- */
    printf("Step 2: Opening dataset for writing...\n");
    rc = amio_open_dataset(core,
                           "examples/manifests/netcdf4_manifest.yaml",
                           AMIO_MODE_WRITE,
                           &dataset);
    check_amio(rc, "amio_open_dataset");
    printf("  Dataset opened.\n");

    /* ---------------------------------------------------------------
     * Step 3: Allocate and fill the temperature array.
     * --------------------------------------------------------------- */
    printf("Step 3: Generating synthetic temperature data...\n");
    float *temperature = (float *)malloc(NX * NY * NZ * sizeof(float));
    if (!temperature) {
        fprintf(stderr, "Failed to allocate temperature array\n");
        return EXIT_FAILURE;
    }
    fill_temperature(temperature, NX, NY, NZ);
    printf("  Data generated: T range ~ [%.1f, %.1f] K\n",
           temperature[0], temperature[NX * NY * NZ - 1]);

    /* ---------------------------------------------------------------
     * Step 4: Write the 3D array through AMIO.
     *
     * We construct an amio_shape_t describing the array layout:
     *   - rank = 3
     *   - extents = {NX, NY, NZ} (row-major: fastest-varying first)
     *   - strides = {0, 0, 0} (contiguous, AMIO derives strides)
     *
     * amio_write() copies the data into the staging pool synchronously,
     * then enqueues the backend write asynchronously. The caller can
     * safely free or reuse the host buffer after this call returns.
     * --------------------------------------------------------------- */
    printf("Step 4: Writing temperature field...\n");
    amio_shape_t shape;
    memset(&shape, 0, sizeof(shape));
    shape.rank = 3;
    shape.extents[0] = NX;
    shape.extents[1] = NY;
    shape.extents[2] = NZ;
    /* strides = 0 means contiguous (AMIO derives row-major strides) */

    rc = amio_write(dataset, "temperature", temperature,
                    AMIO_DTYPE_F32, &shape, &io);
    check_amio(rc, "amio_write(temperature)");
    printf("  Write enqueued (async I/O handle obtained).\n");

    /* ---------------------------------------------------------------
     * Step 5: Flush to ensure all pending writes complete.
     *
     * timeout_ms = 0 means wait indefinitely. In production code,
     * you might set a reasonable timeout (e.g., 30000 ms).
     * --------------------------------------------------------------- */
    printf("Step 5: Flushing pending writes...\n");
    rc = amio_flush(dataset, 0);
    check_amio(rc, "amio_flush");
    printf("  All writes completed successfully.\n");

    /* ---------------------------------------------------------------
     * Step 6: Close the dataset and finalize AMIO.
     *
     * amio_close_dataset() flushes any remaining writes and releases
     * the backend driver resources. amio_finalize() joins worker
     * threads and frees the staging pool.
     * --------------------------------------------------------------- */
    printf("Step 6: Closing dataset and finalizing...\n");
    rc = amio_close_dataset(dataset);
    check_amio(rc, "amio_close_dataset");

    rc = amio_finalize(core);
    check_amio(rc, "amio_finalize");
    printf("  AMIO finalized.\n");

    /* Clean up host memory */
    free(temperature);

    printf("\nDone! Temperature field written to NetCDF-4.\n");
    return EXIT_SUCCESS;
}
